parsers/forth.c

```diff
@@ -0,0 +1,78 @@
+/*
+*   Copyright (c) 2023, Eric Forgeot
+*
+*   Based on work by Jon Strait
+*
+*   This source code is released for free distribution under the terms of the
+*   GNU General Public License version 2 or (at your opinion) any later version.
+*
+*   This module contains functions for generating tags for Forth files
+*/
+
+/*
+*   INCLUDE FILES
+*/
+#include "general.h"	/* must always come first */
+
+#include <ctype.h>
+#include <string.h>
+
+#include "parse.h"
+#include "read.h"
+#include "vstring.h"
+#include "routines.h"
+#include "entry.h"
+
+/*
+*   DATA DEFINITIONS
+*/
+
+typedef enum {
+//   K_SECTION,
+   K_FUNCTION,
+} ForthKind;
+
+static kindDefinition ForthKinds[] = {
+	{ true, 's', "word", "words" },
+};
+
+/*
+*   FUNCTION DEFINITIONS
+*/
+
+static void findForthTags (void)
+{
+	vString *name = vStringNew();
+	const unsigned char *line;
+
+	while ((line = readLineFromInputFile()) != NULL)
+	{
+		if (line[0] == ':') {
+			vStringCatS(name, " ");
+			vStringCatS(name, (const char *) line);
+//			makeSimpleTag(name, K_SECTION);
+			makeSimpleTag(name, K_FUNCTION);
+		}
+		else {
+			vStringClear (name);
+			if (! isspace(*line))
```

Comment:
I don't understand when this branch is needed.
A test case is needed to understand this.

Reply:
it seems to be necessary otherwise it will repeat catching the tags, like this:

![image](https://github.com/universal-ctags/ctags/assets/477243/4621d85e-e024-45c1-94a6-58731c7fc3cf)

Reply:
I don't understand the exmple.

Event with the following change, the test case you prepared is passed.
```
diff --git a/parsers/forth.c b/parsers/forth.c
index d6d7b8cd5..582888d4c 100644
--- a/parsers/forth.c
+++ b/parsers/forth.c
@@ -54,11 +54,7 @@ static void findForthTags (void)
 			vStringCatS(name, (const char *) line);
 //			makeSimpleTag(name, K_SECTION);
 			makeSimpleTag(name, K_FUNCTION);
-		}
-		else {
 			vStringClear (name);
-			if (! isspace(*line))
-				vStringCatS(name, (const char*) line);
 		}
 	}
 	vStringDelete (name);
@@ -77,4 +73,3 @@ extern parserDefinition* ForthParser (void)
 	def->parser = findForthTags;
 	return def;
 }
 ```

Could you prepare a test case that requires the branch?

Reply:
I don't undestand. With the same file example loaded into geany (and without this part), I got this, which is erroneous:

![image](https://github.com/universal-ctags/ctags/assets/477243/a3d80be9-e0e8-45f3-b12f-29c5e66ec365)

Reply:
I don't use Geany. So I don't understand the trouble well.
Could you show the trouble as a test case under Units/simple-forth.d?

Reply:
> I don't use Geany. So I don't understand the trouble well. Could you show the trouble as a test case under Units/simple-forth.d?

Sorry, I didn't read well your diff.  I didn't notice you had kept the ``vStringClear (name);``. Removing it was causing the side effect shown above. Testing this in ctags gave me the same result as in geany.

Using vStringClear in the way you did is correct, and even better it works now if the ``: word ;`` definitions don't have a blank line between them.

I'll update it this way. (sorry, I'm not used to work with branches in git)